File: game/scripts/vscripts/pve/abilities/units/flying_skull/flying_skull_dash/flying_skull_dash.lua
flying_skull_dash = class({})
LinkLuaModifier("modifier_flying_skull_dash_displacement", "pve/abilities/units/flying_skull/flying_skull_dash/modifier_flying_skull_dash_displacement", LUA_MODIFIER_MOTION_BOTH)
LinkLuaModifier("modifier_flying_skull_dash", "pve/abilities/units/flying_skull/flying_skull_dash/modifier_flying_skull_dash", LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_flying_skull_dash_intrinsic", "pve/abilities/units/flying_skull/flying_skull_dash/modifier_flying_skull_dash_intrinsic", LUA_MODIFIER_MOTION_NONE)

function flying_skull_dash:GetIntrinsicModifierName()
	return "modifier_flying_skull_dash_intrinsic"
end

function flying_skull_dash:GetCastPointSpeed()
	return 0
end

function flying_skull_dash:OnAbilityPhaseStart()
	local caster = self:GetCaster()
	local origin = caster:GetAbsOrigin()
    EFX('particles/econ/items/silencer/silencer_ti10_immortal_shield/silencer_ti10_immortal_curse_cast.vpcf', PATTACH_ABSORIGIN, caster, {
        cp0 = origin + Vector(0, 0, 128),
        release = true 
    })
    return true
end

function flying_skull_dash:OnSpellStart()
	local caster = self:GetCaster()
	local origin = caster:GetAbsOrigin()
	local point = self:GetCursorPosition()
    local direction = (point - origin):Normalized()
    local distance = self:GetSpecialValueFor("dash_range")
    local speed = 1200
	self.projectile = nil

	if CustomAbilitiesLegacy:FakeAbility(self) then
		return
	end

    CustomEntitiesLegacy:FullyFaceTowards(caster, direction)
    caster:AddNewModifier(
        caster, -- player source
        self, -- ability source
        "modifier_flying_skull_dash_displacement", -- modifier name
        {
            x = direction.x,
            y = direction.y,
            r = distance,
            speed = speed * RandomFloat(1.0, 1.2),
            peak = 2,
        }
    )

    self.projectile = CustomEntitiesLegacy:ProjectileAttack(caster, {
		tProjectile = {
			EffectName = "particles/vengeful/vengeful_ex_second_attack.vpcf",
			vSpawnOrigin = origin + Vector(direction.x * 45, direction.y * 45, 96),
			fDistance = distance,
			fStartRadius = 1,
			Source = caster,
			vVelocity = direction * speed,
			UnitBehavior = PROJECTILES_DESTROY,
			TreeBehavior = PROJECTILES_NOTHING,
			WallBehavior = PROJECTILES_NOTHING,
			GroundBehavior = PROJECTILES_NOTHING,
			fGroundOffset = 0,
			UnitTest = function(_self, unit) return false end,
		}
	})
end

function flying_skull_dash:GetProjectile()
	return self.projectile
end

function flying_skull_dash:OnUpgrade()
    if IsServer() then
        EFX('particles/flying_skull/burn_main.vpcf', PATTACH_ABSORIGIN_FOLLOW, self:GetCaster(), {
            release = true,
        })
    end
end


if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(flying_skull_dash)

File: src/vscripts/pve/abilities/units/flying_skull_dash.ts
import { CustomAbility } from "../../../abilities/framework/custom_ability";
import { CustomModifier } from "../../../abilities/framework/custom_modifier";
import { registerAbility, registerModifier } from "../../../lib/dota_ts_adapter";
import { DisplacementParams, ModifierDisplacement, OnCollisionEvent } from "../../../modifiers/modifier_displacement";
import { precache, resource } from "../../../precache";
import { direction2D, fullyFaceTowards } from "../../../util";

const resources = precache({
    castFx: resource.fx("particles/econ/items/silencer/silencer_ti10_immortal_shield/silencer_ti10_immortal_curse_cast.vpcf"),
    flyingFx: resource.fx("particles/units/heroes/hero_stormspirit/stormspirit_ball_lightning.vpcf"),
    endFx: resource.fx("particles/units/heroes/hero_vengeful/vengeful_magic_missle_end.vpcf"),
    slowFx: resource.fx("particles/generic_gameplay/rune_haste.vpcf"),
    slowPathFx: resource.fx("particles/econ/items/spectre/spectre_transversant_soul/spectre_transversant_spectral_dagger_path_owner.vpcf"),
    slowStatusFx: resource.fx("particles/status_fx/status_effect_terrorblade_reflection.vpcf")
});

@registerAbility("flying_skull_dash")
class FlyingSkullDash extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_ATTACK;
    }

    GetPlaybackRateOverride() {
        return 0.5;
    }

    GetCastingCrawl() {
        return 0;
    }

    OnAbilityPhaseStart() {
        if (super.OnAbilityPhaseStart()) {
            const origin = this.caster.GetAbsOrigin();
            EFX(resources.castFx.path, ParticleAttachment.ABSORIGIN, this.caster, {
                cp0: origin.__add(Vector(0, 0, 128)),
                release: true
            });
            return true;
        }
        return false;
    }

    OnSpellStart() {
        const point = this.GetCursorPosition();
        const origin = this.caster.GetOrigin();
        const direction = direction2D(origin, point);
        const distance = this.GetSpecialValueFor("dash_range");
        const speed = 1500;
        fullyFaceTowards(this.caster, direction);

        ModifierFlyingSkullDashDisplacement.apply(this.caster, this.caster, this, {
            x: direction.x,
            y: direction.y,
            distance: distance,
            speed: speed * RandomFloat(1.0, 1.2),
            peak: 2,
            radius: 100,
            teamFilter: UnitTargetTeam.ENEMY
        });
    }

    OnUpgrade() {
        if (IsServer()) {
            EFX("particles/flying_skull/burn_main.vpcf", ParticleAttachment.ABSORIGIN_FOLLOW, this.caster, {
                release: true
            });
        }
    }
}

@registerModifier("modifier_flying_skull_dash_displacement")
class ModifierFlyingSkullDashDisplacement extends ModifierDisplacement {
    originalScale!: number;

    OnCreated(params: DisplacementParams) {
        super.OnCreated(params);

        if (IsServer()) {
            this.originalScale = this.parent.GetModelScale();
            this.parent.SetModelScale(this.originalScale / 2);
        }
    }

    OnDestroy() {
        super.OnDestroy();

        if (IsServer()) {
            this.parent.SetModelScale(this.originalScale);
            this.PlayEffectsOnFinish();
        }
    }

    PlayEffectsOnFinish() {
        const position = this.caster.GetAbsOrigin();
        EmitSoundOnLocationWithCaster(position, "Hero_StormSpirit.ProjectileImpact", this.caster);
        EmitSoundOnLocationWithCaster(position, "Hero_StormSpirit.StaticRemnantExplode", this.caster);

        EFX(resources.endFx.path, ParticleAttachment.WORLDORIGIN, undefined, {
            cp0: position,
            cp3: position,
            release: true
        });
    }

    OnCollide(event: OnCollisionEvent) {
        if (IsServer() && event.collision == "unit") {
            for (const unit of event.units) {
                if (!ModifierFlyingSkullDash.findOne(unit)) {
                    ModifierFlyingSkullDash.apply(unit, this.caster, this.ability, { duration: 1.0 });
                    ApplyDamage({
                        victim: unit,
                        attacker: this.parent,
                        damage: this.parent.GetAverageTrueAttackDamage(this.parent),
                        damage_type: DamageTypes.PURE
                    });
                }
            }
        }
    }

    DeclareFunctions() {
        return [ModifierFunction.OVERRIDE_ANIMATION, ModifierFunction.OVERRIDE_ANIMATION_RATE];
    }

    GetOverrideAnimation() {
        return GameActivity.DOTA_RUN;
    }

    GetOverrideAnimationRate() {
        return 1.5;
    }

    GetEffectName() {
        return resources.flyingFx.path;
    }
}

@registerModifier("modifier_flying_skull_dash")
class ModifierFlyingSkullDash extends CustomModifier {
    particleId?: ParticleID;

    GetAttributes() {
        return ModifierAttribute.MULTIPLE;
    }

    OnCreated() {
        if (IsServer()) {
            this.particleId = ParticleManager.CreateParticle(resources.slowPathFx.path, ParticleAttachment.ABSORIGIN_FOLLOW, this.parent);
        }
    }

    OnDestroy() {
        if (IsServer() && this.particleId) {
            ParticleManager.DestroyParticle(this.particleId, true);
            ParticleManager.ReleaseParticleIndex(this.particleId);
        }
    }

    GetEffectName() {
        return resources.slowFx.path;
    }

    DeclareFunctions() {
        return [ModifierFunction.MOVESPEED_BONUS_PERCENTAGE];
    }

    GetModifierMoveSpeedBonus_Percentage() {
        return this.Value("fading_slow_pct");
    }

    GetStatusEffectName() {
        return resources.slowStatusFx.path;
    }

    GetEffectAttachType() {
        return ParticleAttachment.ABSORIGIN_FOLLOW;
    }
}
